File: apps/desktop/electron/sync/websocket-client.ts
import { randomUUID } from 'node:crypto';

import WebSocket from 'ws';

import type { DeviceIdentity } from '../identity/device-identity.js';
import { signDeviceMessage } from '../identity/device-identity.js';

const protocolVersion = '1';

function canonicalHandshake(timestamp: string, nonce: string): string {
  return ['WEBSOCKET', '/api/v1/ws', protocolVersion, timestamp, nonce].join('\n');
}

export class FocusLogWebSocketClient {
  private socket?: WebSocket;
  private heartbeat?: NodeJS.Timeout;
  private reconnect?: NodeJS.Timeout;
  private reconnectAttempt = 0;
  private stopped = true;

  constructor(
    private readonly apiBase: URL,
    private readonly identity: DeviceIdentity,
    private readonly onSyncAvailable: () => void,
    private readonly onRevoked: () => void = () => undefined
  ) {}

  start(): void {
    if (!this.stopped) return;
    this.stopped = false;
    this.connect();
  }

  stop(): void {
    this.stopped = true;
    if (this.heartbeat) clearInterval(this.heartbeat);
    if (this.reconnect) clearTimeout(this.reconnect);
    this.socket?.close(1000, 'client shutdown');
    this.socket = undefined;
  }

  requestReminderClaim(occurrenceId: string, claimTtlSeconds = 30): string {
    return this.send('reminder.claim-request', { occurrenceId, claimTtlSeconds });
  }

  releaseReminderClaim(occurrenceId: string, claimId: string): void {
    this.send('reminder.claim-release', { occurrenceId, claimId });
  }

  private connect(): void {
    if (this.stopped) return;
    const timestamp = new Date().toISOString();
    const nonce = randomUUID();
    const url = new URL('/api/v1/ws', this.apiBase);
    url.protocol = url.protocol === 'https:' ? 'wss:' : 'ws:';
    url.search = new URLSearchParams({
      deviceId: this.identity.deviceId,
      timestamp,
      nonce,
      signature: signDeviceMessage(this.identity, canonicalHandshake(timestamp, nonce)),
      version: protocolVersion
    }).toString();
    const socket = new WebSocket(url);
    this.socket = socket;
    socket.on('open', () => {
      this.reconnectAttempt = 0;
      this.send('presence.hello', {
        state: 'foreground',
        reminderCapable: true,
        clientVersion: 'desktop-0.1.0'
      });
      this.send('sync.subscribe', { cursor: '0' });
    });
    socket.on('message', (data) => this.handle(data.toString()));
    socket.on('close', () => this.scheduleReconnect(socket));
    socket.on('error', () => socket.close());
  }

  private handle(raw: string): void {
    let frame: {
      version?: string;
      type?: string;
      payload?: Record<string, unknown>;
    };
    try {
      frame = JSON.parse(raw) as typeof frame;
    } catch {
      return;
    }
    if (frame.version !== protocolVersion) return;
    if (frame.type === 'connection.ready') {
      const seconds = Number(frame.payload?.heartbeatIntervalSeconds ?? 15);
      if (this.heartbeat) clearInterval(this.heartbeat);
      this.heartbeat = setInterval(
        () => this.send('presence.heartbeat', { state: 'foreground' }),
        Math.max(1, seconds) * 1000
      );
      this.heartbeat.unref();
    } else if (frame.type === 'sync.available') {
      this.onSyncAvailable();
    } else if (frame.type === 'device.revoked') {
      this.onRevoked();
      this.stop();
    }
  }

  private send(type: string, payload: Record<string, unknown>): string {
    const requestId = randomUUID();
    if (this.socket?.readyState === WebSocket.OPEN)
      this.socket.send(
        JSON.stringify({
          version: protocolVersion,
          type,
          requestId,
          sentAt: new Date().toISOString(),
          payload
        })
      );
    return requestId;
  }

  private scheduleReconnect(socket: WebSocket): void {
    if (this.socket !== socket || this.stopped) return;
    if (this.heartbeat) clearInterval(this.heartbeat);
    const exponential = Math.min(30_000, 1_000 * 2 ** this.reconnectAttempt++);
    const delay = Math.round(exponential * (0.8 + Math.random() * 0.4));
    this.reconnect = setTimeout(() => this.connect(), delay);
    this.reconnect.unref();
  }
}

File: apps/mobile/lib/sync/websocket_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:uuid/uuid.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

import '../identity/device_identity.dart';

class FocusLogWebSocketClient {
  FocusLogWebSocketClient({
    required this.endpoint,
    required this.identity,
    required this.onSyncAvailable,
    DeviceIdentityService? identityService,
  }) : _identityService = identityService ?? DeviceIdentityService();

  final Uri endpoint;
  final DeviceIdentity identity;
  final Future<void> Function() onSyncAvailable;
  final DeviceIdentityService _identityService;
  WebSocketChannel? _channel;
  StreamSubscription<dynamic>? _subscription;
  Timer? _heartbeat;
  Timer? _reconnect;
  var _attempt = 0;
  var _stopped = true;
  var _foreground = true;

  Future<void> start() async {
    if (!_stopped) return;
    _stopped = false;
    await _connect();
  }

  Future<void> stop() async {
    _stopped = true;
    _heartbeat?.cancel();
    _reconnect?.cancel();
    await _subscription?.cancel();
    await _channel?.sink.close(1000, 'client shutdown');
    _channel = null;
  }

  void setForeground(bool foreground) {
    _foreground = foreground;
    _send('presence.heartbeat',
        {'state': foreground ? 'foreground' : 'background'});
  }

  String requestReminderClaim(String occurrenceId, {int claimTtlSeconds = 30}) {
    return _send('reminder.claim-request',
        {'occurrenceId': occurrenceId, 'claimTtlSeconds': claimTtlSeconds});
  }

  Future<void> _connect() async {
    if (_stopped) return;
    final timestamp = DateTime.now().toUtc().toIso8601String();
    final nonce = const Uuid().v4();
    const version = '1';
    final signature = await _identityService.sign(identity,
        ['WEBSOCKET', '/api/v1/ws', version, timestamp, nonce].join('\n'));
    final uri = endpoint.replace(
      scheme: endpoint.scheme == 'https' ? 'wss' : 'ws',
      path: '/api/v1/ws',
      queryParameters: {
        'deviceId': identity.deviceId,
        'timestamp': timestamp,
        'nonce': nonce,
        'signature': signature,
        'version': version,
      },
    );
    try {
      final channel = WebSocketChannel.connect(uri);
      await channel.ready;
      if (_stopped) {
        await channel.sink.close();
        return;
      }
      _channel = channel;
      _attempt = 0;
      _subscription = channel.stream.listen(
        _handle,
        onError: (_) => _scheduleReconnect(channel),
        onDone: () => _scheduleReconnect(channel),
        cancelOnError: true,
      );
      _send('presence.hello', {
        'state': _foreground ? 'foreground' : 'background',
        'reminderCapable': true,
        'clientVersion': 'android-0.1.0',
      });
      _send('sync.subscribe', {'cursor': '0'});
    } catch (_) {
      _scheduleReconnect();
    }
  }

  void _handle(dynamic data) {
    try {
      final frame = jsonDecode(data.toString()) as Map<String, dynamic>;
      if (frame['version'] != '1') return;
      final payload = (frame['payload'] as Map<dynamic, dynamic>?)
              ?.cast<String, dynamic>() ??
          const <String, dynamic>{};
      if (frame['type'] == 'connection.ready') {
        final seconds =
            (payload['heartbeatIntervalSeconds'] as num?)?.toInt() ?? 15;
        _heartbeat?.cancel();
        _heartbeat = Timer.periodic(
          Duration(seconds: max(1, seconds)),
          (_) => _send('presence.heartbeat',
              {'state': _foreground ? 'foreground' : 'background'}),
        );
      } else if (frame['type'] == 'sync.available') {
        unawaited(onSyncAvailable());
      } else if (frame['type'] == 'device.revoked') {
        unawaited(stop());
      }
    } on FormatException {
      // Invalid frames are ignored; the server contract remains authoritative.
    }
  }

  String _send(String type, Map<String, dynamic> payload) {
    final requestId = const Uuid().v4();
    _channel?.sink.add(jsonEncode({
      'version': '1',
      'type': type,
      'requestId': requestId,
      'sentAt': DateTime.now().toUtc().toIso8601String(),
      'payload': payload,
    }));
    return requestId;
  }

  void _scheduleReconnect([WebSocketChannel? channelOrNull]) {
    if (_stopped || (channelOrNull != null && _channel != channelOrNull)) {
      return;
    }
    _heartbeat?.cancel();
    _reconnect?.cancel();
    _channel = null;
    final baseSeconds = min(30, 1 << min(_attempt++, 5));
    final jitter = 0.8 + Random().nextDouble() * 0.4;
    _reconnect = Timer(
      Duration(milliseconds: (baseSeconds * 1000 * jitter).round()),
      () => unawaited(_connect()),
    );
  }
}
